File: naivebayeskNNbreastcancer.R
library(e1071)
library(caret)
library(ggpubr)
library(RWeka)
library(ggcorrplot)
setwd("/directory/for/file")
set.seed(0)
cancer_data <- data.frame(read.table("bc_data.csv", header = TRUE, sep = ","))
cancer_for_norm_data <- data.frame(read.table("bc_data.csv", header = TRUE, sep = ","))
names(cancer_data) <- c("id", "diagnosis","radius_mean", "texture_mean", "perimeter_mean","area_mean", "smoothness_mean", "compactness_mean","concavity_mean", "concave_points_mean", "symmetry_mean","fractal_dimension_mean","radius_se","texture_se","perimeter_se","area_se","smoothness_se","compactness_se","concavity_se","concave_points_se","symmetry_se","fractal_dimension_se","radius_worst","texture_worst","perimeter_worst","area_worst","smoothness_worst","compactness_worst","concavity_worst","concave_points_worst","symmetry_worst","fractal_dimension_worst")
names(cancer_for_norm_data) <- c("id", "diagnosis","radius_mean", "texture_mean", "perimeter_mean","area_mean", "smoothness_mean", "compactness_mean","concavity_mean", "concave_points_mean", "symmetry_mean","fractal_dimension_mean","radius_se","texture_se","perimeter_se","area_se","smoothness_se","compactness_se","concavity_se","concave_points_se","symmetry_se","fractal_dimension_se","radius_worst","texture_worst","perimeter_worst","area_worst","smoothness_worst","compactness_worst","concavity_worst","concave_points_worst","symmetry_worst","fractal_dimension_worst")
cancer_data$numerical_diagnosis <- ifelse((cancer_data$diagnosis=="M"),1,0)
#determine correlation this is essential for wanting to determine
#what columns we want to use as predictors for
#a benign and malignant cancer cell
#Visual representation using a correlogram
#correlogram
subset_cancer_data <- subset(cancer_data, select=c(
				numerical_diagnosis,
				radius_mean,
				texture_mean,
				perimeter_mean,
				area_mean,
				smoothness_mean,
				compactness_mean,
				concavity_mean,
				concave_points_mean,
				symmetry_mean,
				fractal_dimension_mean,
				radius_se,
				texture_se,
				perimeter_se,
				area_se,
				smoothness_se,
				compactness_se,
				symmetry_se,
				fractal_dimension_se,
				radius_worst,
				texture_worst,
				perimeter_worst,
				area_worst,
				smoothness_worst,
				compactness_worst,
				concavity_worst,
				concave_points_worst,
				symmetry_worst,
				fractal_dimension_worst
				)
			)
cor(subset_cancer_data)
c_gram_breast_cancer <- round(cor(subset_cancer_data), 1)
# Plots the correlogram
ggcorrplot(
			c_gram_breast_cancer, 
			hc.order = TRUE, 
            type = "lower", 
            lab = TRUE, 
            lab_size = 3, 
            method="circle", 
            colors = c("tomato2", "white", "springgreen3"), 
            title="Correlogram of Gun Violence Data", 
            ggtheme=theme_bw
            )
#sum of columns for normalization
sum_radius_worst <- sum(cancer_for_norm_data$radius_worst) 
sum_concave_points_worst <- sum(cancer_for_norm_data$concave_points_worst)
#sum_concave_points_worst <- sum(cancer_for_norm_data$radius_mean)
sum_area_worst <- sum(cancer_for_norm_data$area_worst)
sum_perimeter_worst <- sum(cancer_for_norm_data$perimeter_worst)
#normalizing the data
cancer_for_norm_data$radius_worst <- cancer_for_norm_data$radius_worst/sum_radius_worst
cancer_for_norm_data$concave_points_worst <- cancer_for_norm_data$concave_points_worst/sum_concave_points_worst
cancer_for_norm_data$area_worst <- cancer_for_norm_data$area_worst/sum_area_worst
cancer_for_norm_data$perimeter_worst <- cancer_for_norm_data$perimeter_worst/sum_perimeter_worst
#Test if normalizing the data changes results and if as factor changes the results
breast_cancer_norm <- subset(cancer_for_norm_data,select = c("id", "diagnosis","radius_worst", "concave_points_worst", "perimeter_worst","area_worst"))
#Naive Bayes Model 0.7427 post worst correlates 0.7661
cancer_data$diagnosis <- as.factor(cancer_data$diagnosis)
cancer_data$radius_worst <- as.factor(cancer_data$radius_worst)
cancer_data$perimeter_worst <- as.factor(cancer_data$perimeter_worst)
cancer_data$area_worst <- as.factor(cancer_data$area_worst)
cancer_data$concave_points_worst <- as.factor(cancer_data$concave_points_worst)
breast_cancer <- subset(cancer_data,select = c("id", "diagnosis","radius_worst", "concave_points_worst", "perimeter_worst","area_worst"))
breast_cancer_factors <- breast_cancer[, 2:6]
breast_cancer_complete <- breast_cancer_factors[complete.cases(breast_cancer_factors),]
breast_cancer_complete$diagnosis <- as.factor(breast_cancer_complete$diagnosis)
data.samples <- sample(1:nrow(breast_cancer_complete),nrow(breast_cancer_complete) * 0.7, replace = FALSE)
training.data <- breast_cancer_complete[data.samples, ]
test.data <- breast_cancer_complete[-data.samples, ]
nb.model <- naiveBayes(diagnosis ~ ., data = training.data)
prediction.nb <- predict(nb.model, test.data)
table(test.data$diagnosis, prediction.nb)
confusionMatrix(prediction.nb,test.data$diagnosis)

#kNN model 10 folds pre norm 0.732865 post norm 0.908612 
# post worst correlates 0.940246
classifier <- IBk(diagnosis ~., data = breast_cancer_norm, control = Weka_control(K = 20, X = TRUE))
evaluate_Weka_classifier(classifier, numFolds = 10)
table(breast_cancer_norm$diagnosis, classifier$predictions)
confusionMatrix(classifier$predictions, breast_cancer_norm$diagnosis)

File: naivebayeskNNbreastcancer.py
import numpy as np
import pandas as pd
import seaborn as sb
import matplotlib.pyplot as plt
import sklearn
from sklearn.naive_bayes import GaussianNB
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import train_test_split

def print_results(predictions,data,desired_variable):
	print("Results:")
	print("Number of mislabeled points out of a total {} points : {}, performance {:05.2f}%"
	.format(
		test_data.shape[0],
		(data[desired_variable] != predictions).sum(),
		100*(1-(data[desired_variable] != predictions).sum()/data.shape[0])
		)
	)

def plot_correlations(data,correl_columns):
	correlations = pd.DataFrame()
	for main_variable in correl_columns:
		for other_variable in correl_columns:
			correlations.loc[main_variable, other_variable] = data.corr().loc[main_variable, other_variable]
	sb.heatmap(correlations)
	plt.show()
	return correlations

def normalize_columns(data, normalize_columns):
	for column in normalize_columns:
		sum_of_column = data[column].sum()
		data[column] = data[column]/sum_of_column
	return data

np.random.seed(0)
cancer_data = pd.read_csv("bc_data.csv")
cancer_data['numerical_diagnosis'] = np.where(cancer_data['diagnosis']=='M', 1, 0)
cancer_for_norm_data = cancer_data
correl_columns = ['numerical_diagnosis','radius_mean','texture_mean','perimeter_mean','area_mean','smoothness_mean','compactness_mean','concavity_mean','concave_points_mean','symmetry_mean','fractal_dimension_mean','radius_se','texture_se','perimeter_se','area_se','smoothness_se','compactness_se','symmetry_se','fractal_dimension_se','radius_worst','texture_worst','perimeter_worst','area_worst','smoothness_worst','compactness_worst','concavity_worst','concave_points_worst','symmetry_worst','fractal_dimension_worst']
cancer_data_correl = cancer_data[correl_columns]
correlations = plot_correlations(cancer_data_correl,correl_columns)
columns_to_normalize = ["radius_worst","concave_points_worst","area_worst","perimeter_worst"]
cancer_for_norm_data = normalize_columns(cancer_for_norm_data, columns_to_normalize)

#Naive Bayes
#subset to change our cancer_data to train our statistical model
factors_list = ["numerical_diagnosis","radius_worst","perimeter_worst","area_worst","concave_points_worst"]
predictor_factors = ["radius_worst","perimeter_worst","area_worst","concave_points_worst"]
breast_cancer_factors = cancer_data[factors_list]
breast_cancer_complete = breast_cancer_factors.dropna()
data_sample = breast_cancer_complete.sample(frac=0.7, replace=False)
training_data = data_sample
test_data = breast_cancer_complete[~breast_cancer_complete.isin(training_data)].dropna()

nb_model = GaussianNB()
nb_model.fit(training_data[predictor_factors],training_data["numerical_diagnosis"])

test_pred = nb_model.predict(test_data[predictor_factors])
print_results(test_pred,test_data,"numerical_diagnosis")
#93.57 % test
#94.33 % train


#kNN
#subset to change our cancer_data to train our statistical model
#id diagnosis radius worst perimeter worst area worst concave points worst
#Used 10 k nearest neighbors for classification 
breast_cancer_norm = cancer_for_norm_data[factors_list]
data_sample = breast_cancer_norm.sample(frac=0.7, replace=False)
training_data = data_sample
test_data = breast_cancer_norm[~breast_cancer_norm.isin(training_data)].dropna()
knn_model = KNeighborsClassifier(n_neighbors=15)
knn_model.fit(training_data[predictor_factors],training_data["numerical_diagnosis"])
test_pred = knn_model.predict(test_data[predictor_factors])
print_results(test_pred,test_data,"numerical_diagnosis")

#95.32 test data k = 10 95.91 k = 15
